dis_snek/models/discord/channel.py

```diff
@@ -776,13 +778,53 @@ def category(self) -> Optional["GuildCategory"]:
 
     @classmethod
     def _process_dict(cls, data: Dict[str, Any], client: "Snake") -> Dict[str, Any]:
-        data["original_permission_overwrites"] = data.get("permission_overwrites", [])
-        data["permission_overwrites"] = {
-            obj.id: obj
-            for obj in (PermissionOverwrite(**permission) for permission in data["original_permission_overwrites"])
-        }
+        if overwrites := data.get("permission_overwrites"):
+            data["permission_overwrites"] = [PermissionOverwrite.from_dict(overwrite) for overwrite in overwrites]
         return data
 
+    def permissions_for(self, instance: Snowflake_Type) -> Permissions:
+        """
+        Calculates permissions for an instance
+
+        Args:
+            instance: Member or Role instance (or its ID)
+
+        Returns:
+            Permissions data
+
+        Raises:
+            ValueError: If could not find any member or role by given ID
+            RuntimeError: If given instance is from another guild
+
+        """
+        if (is_member := isinstance(instance, models.Member)) or isinstance(instance, models.Role):
+            if instance._guild_id != self._guild_id:
+                raise RuntimeError("Unable to calculate permissions for the instance from different guild")
+
+            if is_member:
+                return instance.channel_permissions(self)
+
+            else:
+                permissions = instance.permissions
+
+                for overwrite in self.permission_overwrites:
+                    if overwrite.id == instance.id:
+                        permissions &= ~overwrite.deny
+                        permissions |= overwrite.allow
+                        break
+
+                return permissions
+
+        else:
+            instance = to_snowflake(instance)
+            instance = self._client.cache.get_member(self._guild_id, instance) or self._client.cache.get_role(instance)
+            # Will be replaced with `Guild.get_` just after #289 merge!
```

Comment:
Do not merge this PR unless its done!

Reply:
![](https://cdn.discordapp.com/emojis/908079211800051712.webp?size=96&quality=lossless)

Reply:
Done